glass/grf/_core.py

```diff
@@ -0,0 +1,219 @@
+from __future__ import annotations
+
+import functools
+from typing import TYPE_CHECKING, Protocol, TypeVar
+
+import transformcl
+
+if TYPE_CHECKING:
+    from collections.abc import Callable
+
+
+ArrayT = TypeVar("ArrayT", bound="Array")
+ArrayT_co = TypeVar("ArrayT_co", bound="Array", covariant=True)
+TransformationT = TypeVar("TransformationT", bound="Transformation")
+
+
+class ArrayNamespace(Protocol[ArrayT]):
+    """Protocol for array namespaces."""
+
+    pi: float
+
+    def arange(self, n: int) -> ArrayT: ...
+
+    def sqrt(self, x: ArrayT) -> ArrayT: ...
+    def exp(self, x: ArrayT) -> ArrayT: ...
+    def expm1(self, x: ArrayT) -> ArrayT: ...
+    def log1p(self, x: ArrayT) -> ArrayT: ...
+
+
+class Array(Protocol):
```

Comment:
I don't see any classes structurally sub-typing `Array` in this PR. Is this required or am I missing something? Can the arguments of type `ArrayT` below be just `npt.NDArray[...]`?

I feel like this is in preparation of Array API, but I would prefer to keep it in a separate PR.

Reply:
Ok, this can come later. I added it here because it should already be fully compliant. But I can keep it NDArray here and then we change everything in one go.